README.md
K Phone Book
=========================
## Descripton
Simple Phone book application
CRUD operations for in build json data
Operations performed in modal popup
Validation provided for form elements using ng-form validation.
Application can be started at http//localhost:7000 using glup serve at command prompt

Set up
======

1. Clone this git repo. Run `npm install` to download `devDependencies`.

2.  Personalize the following files for your project:
    - `LICENSE`
    - `package.json`
    - `CHANGELOG.md`
    - `README.md`

    See [npm docs](https://npmjs.org/doc/json.html) for more on customizing `package.json`.


## gulp

gulpfile.js is used for starting gulp. 
Coffee script has been used to register tasks in gulpfile.js

Customize `gulpfile.js` as necessary. See [gulp docs](https://github.com/gulpjs/gulp).


Development Workflow
====================

Open Command prompt then:

1.  Run `npm install`

2.  Run `bower install`

3.  Copy bower_components folder to app/
    
4.  Run `gulp serve`

5.  Enter http://localhost:7000 in browser (Chrome) to start the phone book web application




CHANGELOG.md
cofeescript-seed-project changelog
==================================

Edit `CHANGELOG.md` for your own project.
Using update date instead of semver convention for coffeescript-seed-project.


## Feb 18, 2014

- Update `package.json`.
    - Update `devDependencies`.
    - Add `engines` property.
    - Remove webpack task
- Add `CHANGELOG.md`.
- Update `mocha.opts` to support CoffeeScript 1.7+.
- Update `README.md` to reflect changes.
    - Add david-dm badge (watch dependencies)
- Use `through2` in place of `through`.
- Use webpack within `gulpile.js`.
- Use CoffeeScript files as primary source files.
- Update `test/test.coffee`.
- Replace `gulp-coffee` with webpack plugin `coffee-loader`.
- Generate dev source map file with webpack.
- Generate dist source map file with gulp-uglify.
- Remove `gulp-if`.
- Added mocha test automation task within `gulpfile.js`.


app/app.js
'use strict';

/**
 * @ngdoc overview
 * @name kphoneBookApp
 * @description
 * # kphoneBookApp
 *
 * Main module of the application.
 */
angular
  .module('kphoneBookApp', [
    'ngAnimate',
    'ngCookies',
    'ngResource',
    'ui.router',
    'ngSanitize',
    'ngTouch',
    'ui.bootstrap',
    'messages.factory'
  ]).config(['$stateProvider', '$urlRouterProvider', function($stateProvider, $urlRouterProvider) {
    $urlRouterProvider.otherwise('/home');
        $stateProvider.state('home', {
            url: '/home',
            params: {},
            views: {
                'main@': {
                    controller: 'HomeCtrl',
                    templateUrl: 'home.html'
                }
            },
            data: {
                pageMainTitle: 'K-Phone Book Home',
                displayName: 'K-Phone Book Home',
                pageSubTitle: null
            }
        })
        .state('header', {
            url: '/header',
            params: {},
            views: {
                'header@': {
                    templateUrl: 'header.html'
                }
            },
            data: {
                pageMainTitle: 'K-Phone Book',
                displayName: 'K-Phone Book',
                pageSubTitle: null
            }
        })
        .state('phonebook', {
            url: '/phonebook',
            params: {},
            views: {
                'main@': {
                    controller: 'PhonebookCtrl',
                    templateUrl: 'kphoneBook.html'
                }
            },
            data: {
                pageMainTitle: 'K-Phone Book',
                displayName: 'K-Phone Book',
                pageSubTitle: null
            }
        });
}]);


app/phonebook.controller.js
/**
 * @ngdoc function
 * @name ktoolsApp.controller:AboutCtrl
 * @description
 * # PhonebookCtrl
 * Controller of the kphoneBookApp
 */
angular.module('kphoneBookApp')
  .controller('PhonebookCtrl', function ($scope, $uibModal, Messages, $rootScope) {
  	$scope.messages = new Messages();
  	$rootScope.alerts = [];
  	$rootScope.uContact = {};
    $rootScope.contacts =
		[
			{firstName:'Dave', lastName:'Jameson', number:0546655837},
			{firstName:'Larisa', lastName:'Dorfman', number:050923423},
			{firstName:'John', lastName:'Oxford', number:0529346353},
			{firstName:'Joqqhn', lastName:'Oxford', number:0522346353}
		];
	$scope.isUpdate = false;
	$scope.isView = false;
	$scope.selectedContact = {};
	$scope.addBtnLabel = 'Add';
	$scope.$on('record::add', function(event, eventData) {
		$rootScope.contacts.push({firstName:$scope.firstName,
		lastName:$scope.lastName, number:$scope.number});
		frmPhoneBook.txtFirstName.value = '';
		frmPhoneBook.txtLastName.value = '';
		frmPhoneBook.txtNumber.value = '';
	});
	$scope.addContact = function() {
		if(frmPhoneBook.txtFirstName.value && frmPhoneBook.txtLastName.value && frmPhoneBook.txtNumber.value) {
	        $scope.contacts.push({firstName:frmPhoneBook.txtFirstName.value,
			lastName:frmPhoneBook.txtLastName.value, number:frmPhoneBook.txtNumber.value});
			frmPhoneBook.txtFirstName.value = '';
			frmPhoneBook.txtLastName.value = '';
			frmPhoneBook.txtNumber.value = '';
		}
	};
	$scope.modalOpen = function(updateMode, selContact) {
		$rootScope.selectedContact = {};
		$rootScope.alerts = [];
		if (updateMode === 'Update') {
			$scope.isUpdate = true;
			$scope.isView = false;
			$rootScope.contactToUpdate = angular.copy(selContact);
			$rootScope.uContact.firstName = selContact.firstName;
			$rootScope.uContact.lastName = selContact.lastName;
			$rootScope.uContact.number = selContact.number;
		}else if (updateMode === 'View') {
			$scope.isUpdate = false;
			$scope.isView = true;
			$rootScope.selectedContact.firstName= selContact.firstName;
			$rootScope.selectedContact.lastName = selContact.lastName;
			$rootScope.selectedContact.number = selContact.number;
		}else {
			$rootScope.uContact = {};
		}
		$scope.modalInstance = $uibModal.open({
			templateUrl: 'updatePhonebook.html',
			controller: /*@ngInject*/ function($scope, $uibModalInstance, $rootScope) {
				$scope.addBtnLabel = updateMode;
				if (updateMode === 'View') {
					$scope.isUpdate = false;
					$scope.isView = true;
				}else{
					$scope.isUpdate = true;
					$scope.isView = false;
				}
				$scope.modalYes = function($scope) {
					if (updateMode === 'Add') {
						$rootScope.contacts.push({firstName:$rootScope.uContact.firstName,
						lastName:$rootScope.uContact.lastName, number:$rootScope.uContact.number});
						$rootScope.alerts = [];
						$rootScope.alerts.push({ type: 'success', msg: 'New contact [' + $rootScope.uContact.firstName + '] created successfully' });
						$rootScope.uContact = {};
					} else if (updateMode === 'Update') {
						angular.forEach($rootScope.contacts, function(eachContact) {
							if (eachContact.firstName === $rootScope.contactToUpdate.firstName) {
								eachContact.firstName = $rootScope.uContact.firstName;
								eachContact.lastName = $rootScope.uContact.lastName;
								eachContact.number = $rootScope.uContact.number;
								$rootScope.alerts = [];
								$rootScope.alerts.push({ type: 'success', msg: 'Contact [' + $rootScope.uContact.firstName + '] updated successfully' });
								$rootScope.uContact = {};
							}
						});
					}
					$uibModalInstance.close('Yes');
				};
				$scope.modalNo = function() {
					$uibModalInstance.close('No');
				};
				$scope.modalCancel = function() {
					$uibModalInstance.dismiss('Cancel');
				};
			},
			size: '300',
			backdrop: 'static',
			animation: true,
			resolve: {}
		});
	};
	$scope.deleteContact = function(selContact) {
		$scope.messages.confirm('Are you sure to delete the contact [' + selContact.firstName + ']', function(confirm) {
			if (confirm) {
				$rootScope.contacts.splice($rootScope.contacts.indexOf(selContact), 1);	
				$rootScope.alerts = [];
				$rootScope.alerts.push({ type: 'success', msg: 'Contact [' + $rootScope.uContact.firstName + '] updated successfully' });
			}	
		});
	};
	
	$rootScope.closeAlert = function(index) {
	    $rootScope.alerts.splice(index, 1);
	};
  });


app/factory/messages.factory.js
(function() {
    'use strict';

    /**
     * @ngdoc Factory for the messages model object
     *
     * @description The messages model object to encapsulate loading, information, warning,
     * and error messages
     */
    angular.module('messages.factory', []).factory('Messages', function(
        $filter, $uibModal, $log, $timeout) {

        function Messages() {
            /* Data fields */
            this.loading = null;
            this.error = null;
            this.success = null;
            this.isDownload = null;

            // Scroll settings
            this.scroll = false;

            // Fadeout settings
            this.fadeoutInfo = {};

            /* Chainable set methods */
            this.showLoading = function(msg) {
                var loadingMsg = $filter('translate')('loading');
                loadingMsg += '...';
                this.loading = isPrintable(msg) ? msg : loadingMsg;
                return this;
            };

            this.showError = function(msg, scroll) {
                this.error = isPrintable(msg) ? msg : 'An error has occurred.';
                this.scroll = angular.isUndefined(scroll) ? true : scroll;
                return this;
            };

            this.showSucess = function(msg, scroll) {
                this.success = isPrintable(msg) ? msg : 'Operation successful!';
                this.scroll = angular.isUndefined(scroll) ? true : scroll;

                var that = this;
                this.fadeout(function() {
                    that.success = null;
                });

                return this;
            };

            this.showWarning = function(msg, scroll) {
                this.warning = isPrintable(msg) ? msg : 'Warning';
                this.scroll = angular.isUndefined(scroll) ? true : scroll;
                return this;
            };

            this.clear = function() {
                this.loading = null;
                this.error = null;
                this.success = null;
                this.warning = null;
                return this;
            };

            // Shows an alert modal, expects the following arguments
            // alert(message, callbackFn) or alert(options) - @see showModal function for options
            this.alert = function alert() {
                showModal(parseModalArgs(arguments), false);
            };

            // Shows an confirm modal, expects the following arguments
            // confirm(message, callbackFn) or confirm(options) - @see showModal function for options
            this.confirm = function confirm() {
                showModal(parseModalArgs(arguments), true);
            };

            // Fadeout
            this.fadeout = function fadeout(callback) {
                var FADE_TIMEOUT_MS = 8000;

                // If there is an existing timer, cancel it
                $timeout.cancel(this.fadeoutInfo.timer);

                // Start a new timer for the new message
                this.fadeoutInfo.timer = $timeout(callback, FADE_TIMEOUT_MS);
            };
        }

        return Messages;

        // Returns true if the input message is printable
        function isPrintable(msg) {
            return angular.isDefined(msg) &&
                   (angular.isString(msg) ||
                    angular.isNumber(msg) ||
                    angular.isDate(msg));
        }

        // Opens up a Modal window for alerts and confirmations
        // severity "info" (default), "success", "warning", "error"
        function showModal(options, confirm) {
            options.confirm = confirm;
            options.severity = options.severity ? options.severity : "info";
            options.title = options.title || getDefaultTitle(confirm, options.severity);
            options.message = options.message || ' ';
            options.okLabel = options.okLabel || (confirm ? 'Yes' : 'OK');
            options.cancelLabel = options.cancelLabel || 'No';
            options.callbackFn = options.callbackFn || angular.noop;
            options.size = options.size || 'md';

            // TODO: Throw error if message and callback are not set!

            $uibModal.open({
                animation: true,
                templateUrl: 'templates/confirm.modal.template.html',
                controller: /*@ngInject*/ function($scope, options, $uibModalInstance) {
                    var vm = this;
                    vm.options = options;
                    vm.ok = function() {
                        vm.options.callbackFn(true);
                        $uibModalInstance.close();
                    };
                    vm.cancel = function() {
                        vm.options.callbackFn(false);
                        $uibModalInstance.close();
                    };
                    vm.close = function() {
                        try {
                            if (vm.options.confirm) {
                                vm.cancel();
                            } else {
                                vm.ok();
                            }
                        } catch (e) {
                            $log.log('Error closing via ok or cancel', e);
                            $uibModalInstance.close();
                        }
                    };
                },
                controllerAs: 'vm',
                backdrop: 'static',
                size: options.size,
                windowClass: "alert-confirm-window " + options.severity,
                windowTopClass: "alert-confirm-windowtop",
                backdropClass: "alert-confirm-backdrop",
                resolve: {
                    options: function() {
                        return options;
                    }
                }
            });

            function getDefaultTitle(confirm, severity) {
                if (confirm) {
                    return "confirmation";
                } else { // Alert
                    switch (severity) {
                        case "info":
                            return 'information';
                        case "success":
                            return 'success';
                        case "warning":
                            return 'warning';
                        case "error":
                            return 'error';
                        default:
                            return 'alert';
                    }
                }
            }
        }

        function parseModalArgs(args) {
            var options = {};
            switch (args.length) {
                case 1:
                    options = angular.isObject(args[0]) ?
                              args[0] : {message: args[0]};
                    break;
                case 2:
                    options = {message: args[0],
                               callbackFn: args[1]};
                    break;
                default:
                    throw "Unknown number of arguments, should be either " +
                          "[1 - options] or " +
                          "[2 - message, callback]";
            }
            return options;
        }
    });
})();


gulpfile.js
/*jshint node:true */
/* gulpfile.js - https://github.com/gulpjs/gulp */

// for mocha
require('coffee-script/register');

var
/** node.js **/
path = require('path'),
fs = require('fs'),

/** Gulp and plugins **/
gulp = require('gulp'),
gutil = require('gulp-util'),
watch = require('gulp-watch'),
plumber = require('gulp-plumber'),
uglify = require('gulp-uglify'),
rename = require('gulp-rename'),
mocha = require('gulp-mocha'),
connect = require('gulp-connect'),
coffee = require('gulp-coffee');

/** utility **/
through = require('through2').obj,

/** webpack **/
webpack = require('webpack'),
webpackconfig = require('./webpack.config'),

/** Config **/
distDir = './dist/',
distTargetFile = 'myawesomeproject.js',

/** Environment Vars **/
R = 0,
ENV_SWITCH = void 0,

// Env list
DEV_ENV = R++,
PROD_ENV = R++;


/** Utility Functions **/

// Transformation function on gulp.src depending on env
var getGlob = function(glob_target, opts) {
  var src = gulp.src(glob_target, opts);

  switch(ENV_SWITCH) {
    case DEV_ENV:

      // watch files and re-emit them downstream on change (or some file event)
      opts.glob = glob_target;

      return watch(opts)
                .pipe(plumber())
                .pipe(through());

    case PROD_ENV:
      return src.pipe(plumber());

    default:
      throw new Error('Invalid Env');
  }
};

/* Sub-tasks */
gulp.task('set-dev', function() {
    ENV_SWITCH = DEV_ENV;
});

gulp.task('set-prod', function() {
    ENV_SWITCH = PROD_ENV;
});

gulp.task('serve', function() {
  connect.server({
    root: ['app', 'tmp'],
    port: 7000,
    hostname: 'vrkotesh',
    livereload: true
  });
});
 
gulp.task('livereload', function() {
  gulp.src(['.tmp/styles/*.css', '.tmp/scripts/*.js'])
    .pipe(watch())
    .pipe(connect.reload());
});

gulp.task('coffee', function() {
  gulp.src('src/*.coffee')
    .pipe(coffee())
    .pipe(gulp.dest('.tmp/scripts'));
})

gulp.task('connectDev', function () {
  connect.server({
    root: ['app', 'tmp'],
    port: 8000,
    hostname: 'vrkotesh',
    livereload: true
   });
});

gulp.task('mocha', function() {

    var mocha_opts = {};

    try {
        var opts = fs.readFileSync('test/mocha.opts', 'utf8')
            .trim()
            .split(/\s+/);

        opts.forEach(function(val, indx, arry) {
            if (/^-.+?/.test(val)) {
                val = val.replace(/^-+(.+?)/, "$1");
                mocha_opts[val] = arry[indx + 1];
            }
        });

    } catch (err) {
      // ignore
    }

    return watch({ glob: 'test/**/*.coffee', read:false }, function(files) {

        files
            .pipe(mocha(mocha_opts))
                .on('error', function(err) {
                    if (!/tests? failed/.test(err.stack)) {
                        console.log(err.stack);
                    }
                });
    });

});

gulp.task('dist-minify', function() {

    return getGlob(distDir + '/' + distTargetFile)
        .pipe(rename({ suffix: '.min'}))
        .pipe(uglify({ outSourceMap: true }))
        .pipe(gulp.dest(distDir));
});

gulp.task('webpack', function(callback) {

    webpack(webpackconfig, function(err, stats) {

        if(err) throw new gutil.PluginError("webpack", err);

        gutil.log("[webpack]", stats.toString({
            // output options
        }));

    });

    callback();

});

/* High-level tasks */
/* Compose sub-tasks to orchestrate something to be done */

/* Development task */
gulp.task('dev', ['set-dev', 'webpack', 'mocha'], function() {
    // Run webpack based on config from webpack.config.js
});


gulp.task('prod', ['set-prod', 'dist-minify'], function() {
    // Minify file and generate dist source map file.
});


// The default task (called when you run `gulp`)
gulp.task('default', ['dev'], function() {
    // Run dev task by default
});
